conctr.agent.lib.nut

```diff
@@ -195,18 +232,296 @@ class Conctr {
+    // @param  {String/Array}   deviceId     Device id(s) the message should be sent to
+    // @param  {[type]}         msg          Data to be sent to be published
+    // @param  {[type]}         contentType  Header specifying the content type of the msg
+    // @param  {Function}       cb           Function called on completion of publish request
+    // 
+    function publishToDevice(deviceIds, msg, contentType = null, cb = null) {
+        local relativeUrl = "/dev/";
+        _publish(relativeUrl, { "device_ids": deviceIds }, msg, contentType, cb);
+    }
+
+
+    // 
+    // Subscribe to a single/list of topics
+    // 
+    // @param  {Function}       cb     Function called on receipt of data
+    // @param  {Array/String}   topics String or Array of string topic names to subscribe to
+    // 
+    function subscribe(topics = [], cb = null) {
+
+        // Check if only callback is subscribed to
+        if (typeof topics == "function") {
+            cb = topics;
+            topics = [];
+        }
+
+        if (cb == null) {
+            throw "Conctr: callback to subscribe is a required parameter";
+        }
+
+        // If no topics to subscribe to found, subscribe to this devices id
+        if (topics.len() == 0) {
+            topics = ["dev/" + _device_id];
+        }
+
+        if (typeof topics != "array") {
+            topics = [topics];
+        }
+
+        local action = "subscribe";
+        local headers = {};
+        local payload = {};
+        local chunks = "";
+        local contentLength = null;
+        local reqTime = time();
+
+        // http done callback
+        local _doneCb = function(resp) {
+
+            // We dont allow non chunked requests. So if we recieve a message in this func
+            // it is the last message of the steam and may contain the last chunk
+            if (resp.body == null && resp.body == "") {
+                _streamCb(resp.body);
+            }
+
+            local wakeupTime = 0;
+            local reconnect = function() {
+                subscribe(topics, cb);
+            }
+
+            if (resp.statuscode >= 200 && resp.statuscode <= 300) {
+                // wake up time is 0
+            } else if (resp.statuscode == 429) {
+                wakeupTime = 1;
+            } else if (resp.statuscode == 401) {
+                throw "Conctr: Authentication failed";
+            } else {
+                local conTime = time() - reqTime;
+                if (conTime < MIN_RECONNECT_TIME) {
+                    wakeupTime = MIN_RECONNECT_TIME - conTime;
+                }
+                server.error("Conctr: Subscribe failed with error code " + resp.statuscode + ". Retrying in " + wakeupTime + " seconds");
+            }
+
+            // Reconnect in a bit or now based on disconnection reason
+            imp.wakeup(wakeupTime, reconnect.bindenv(this));
+        };
+
+        // Http streaming callback
+        local _streamCb = function(chunk) {
```

Comment:
Same as for _doneCb. It would be a lot more readable if this function wasn't declared within the subscribe function but was a private method on the class.